new readme and documentation (#106)
kioku-project/kioku: doc/backend_development.md

# Backend development

## Index
1. [Infrastructure overview](#infrastructure-overview)
2. [Getting started](#getting-started-with-backend-development)
3. [Create a new service](#create-a-new-service)

## Infrastructure overview
![Architecture-Design](https://github.com/kioku-project/kioku/assets/60541979/4aedb2be-f1ff-41bc-ac7f-a45662bd819a)

> The backend is designed with modern microservice philosophy in mind. The communication between the microservices themselves is done using gRPC which allows for fast and streamlined communication. All services can only be accessed from outside through the frontend_proxy service which translates rest api calls to the corresponding grpc methods. The frontend_proxy is also responsible for the authentication of the user, allowing the services to assume the provided userID as save to use. The services are written in the [go](https://go.dev/) programming language and use [go-micro](https://github.com/go-micro/go-micro)https://github.com/go-micro/go-micro as a framework.

Reviewer: ```suggestion
The backend is designed with modern microservice philosophy in mind. The communication between the microservices themselves is done using gRPC which allows for fast and streamlined communication. All services can only be accessed from outside through the `frontend_proxy` service which translates rest api calls to the corresponding gRPC methods. The `frontend_proxy` is also responsible for the authentication of the user, allowing the services to assume the provided userID as save to use. The services are written in the [go](https://go.dev/) programming language and use [go-micro](https://github.com/go-micro/go-micro)https://github.com/go-micro/go-micro as a framework.
```

Author: implemented in ee4f7a25

Reviewer: The double link for go-micro still exists

Author: oh sorry, I overlooked that. Fixed in 0dafe9e4

> The backend is designed with modern microservice philosophy in mind. The communication between the microservices themselves is done using gRPC which allows for fast and streamlined communication. All services can only be accessed from outside through the frontend_proxy service which translates rest api calls to the corresponding grpc methods. The frontend_proxy is also responsible for the authentication of the user, allowing the services to assume the provided userID as save to use. The services are written in the [go](https://go.dev/) programming language and use [go-micro](https://github.com/go-micro/go-micro)https://github.com/go-micro/go-micro as a framework.

Reviewer 2: ```suggestion
The backend is designed with modern microservice philosophy in mind. The communication between the microservices themselves is done using gRPC which allows for fast and streamlined communication. All services can only be accessed from outside through the frontend_proxy service which translates REST API calls to the corresponding grpc methods. The frontend_proxy is also responsible for the authentication of the user, allowing the services to assume the provided userID as safe to use. The services are written in the [go](https://go.dev/) programming language and use [go-micro](https://github.com/go-micro/go-micro)https://github.com/go-micro/go-micro as a framework.
```

Author: implemented in ee4f7a25

## Getting started with backend development

> Before getting started, ensure that you have a recent version of golang installed. Version 1.16 or higher is required.

Reviewer 2: ```suggestion
Before getting started, ensure that you have a recent version of [Go](https://golang.org) installed. Version 1.16 or higher is required.
```

Author: implemented in ee4f7a25

```
go install github.com/kioku-project/go-micro-cli/cmd/go-micro@latest
```
Additionally, you need to set up the `protoc` compiler in your environment. See [here](https://grpc.io/docs/protoc-installation/) for installation instructions according to your system.
And finally, you need to install `make` on your system. Mac users can install it alongside common developer packages via `xcode-select --install`, Windows users are able to obtain it via [Chocolatey](https://chocolatey.org/install) by using the command `choco install make`

## Create a new service
1. Set up a new service

        cd backend/services/
        go-micro new service --health --kubernetes github.com/kioku-project/kioku/services/<name-of-new-service>
        cd <name-of-new-service>
        make init proto update tidy

2. Make adjustments
    1. Update the service definitions in `proto/<name-of-new-service>.proto`
    2. Generate proto files by running `make proto`
    3. Update handler in `handler/<name-of-new-service>.go`

3. Update docker-compose
    
    Add a new service in `docker-compose.yml` for the created service.

> 

Reviewer 3: Maybe it would also be beneficial to document how to add the new service to the Kubernetes deployment

Author: implemented in ee4f7a25

Reviewer 3: I cannot find this addition

Author: ![image](https://github.com/kioku-project/kioku/assets/60541979/d59372ff-97ee-415b-ad81-bfa39d8bb2ca)

Reviewer 3: Sorry for the misunderstanding. I meant the Kubernetes deployment, not the docker-compose. If you want I can write this

Author: Oh, then I misunderstood your request, sorry.
Yes, it would be very nice if you could add this or write a suggestion.

# Backend development

## Index
1. [Infrastructure overview](#infrastructure-overview)
2. [Getting started](#getting-started-with-backend-development)
3. [Create a new service](#create-a-new-service)

## Infrastructure overview
![Architecture-Design](https://github.com/kioku-project/kioku/assets/60541979/4aedb2be-f1ff-41bc-ac7f-a45662bd819a)

> The backend is designed with modern microservice philosophy in mind. The communication between the microservices themselves is done using gRPC which allows for fast and streamlined communication. All services can only be accessed from outside through the `frontend_proxy` service which translates REST API calls to the corresponding gRPC methods. The `frontend_proxy` is also responsible for the authentication of the user, allowing the services to assume the provided userID as safe to use. The services are written in the [go](https://go.dev/) programming language and use [go-micro](https://github.com/go-micro/go-micro) as a framework.

Reviewer 2: ```suggestion
The backend is designed with modern microservice philosophy in mind. The communication between the microservices themselves is done using gRPC which allows for fast and streamlined communication. All services can only be accessed from outside through the `frontend_proxy` service which translates REST API calls to the corresponding gRPC methods. The `frontend_proxy` is also responsible for the authentication of the user, allowing the services to assume the provided userID as safe to use. The services are written in the [Go](https://go.dev/) programming language and use [go-micro](https://github.com/go-micro/go-micro) as a framework.
```

## Getting started with backend development

> Before getting started, ensure that you have a recent version of [Go](https://golang.org) installed. Version 1.16 or higher is required.

Reviewer 2: ```suggestion
Before getting started, ensure that you have a recent version of [Go](https://go.dev) installed. Version 1.16 or higher is required.
```

And finally, you need to install `make` on your system. Mac users can install it alongside common developer packages via `xcode-select --install`, Windows users are able to obtain it via [Chocolatey](https://chocolatey.org/install) by using the command `choco install make`

## Create a new service
1. Set up a new service

        cd backend/services/
        go-micro new service --health --kubernetes github.com/kioku-project/kioku/services/<name-of-new-service>
        cd <name-of-new-service>
        make init proto update tidy

2. Make adjustments
    1. Update the service definitions in `proto/<name-of-new-service>.proto`
    2. Generate proto files by running `make proto`
    3. Update handler in `handler/<name-of-new-service>.go`

3. Update docker-compose
    
    Add a new service in `docker-compose.yml` for the created service.
    ```yaml
<name>_service:
  build:
    context: backend
    dockerfile: services/<name>/Dockerfile
  container_name: kioku-<name>_service
  restart: always
  env_file:
    - ./backend/.env
  depends_on:
    - db

>     ```

Reviewer 3: This YAML is not being rednered in the README. Maybe an indentation problem?

>     ```yaml
> <name>_service:
>   build:
>     context: backend
>     dockerfile: services/<name>/Dockerfile
>   container_name: kioku-<name>_service
>   restart: always
>   env_file:
>     - ./backend/.env
>   depends_on:
>     - db
>     ```
> 
> 4. Adjust the proxy rules in the frontend service to be able to serve the new service if needed

Reviewer 3: ```suggestion
    ```yaml
    <name>_service:
      build:
        context: backend
        dockerfile: services/<name>/Dockerfile
      container_name: kioku-<name>_service
      restart: always
      env_file:
        - ./backend/.env
      depends_on:
        - db
    ```

4. Add Service to GitHub Workflows in `.github/workflows/<name>_service.yaml`
```yaml
name: <name> Service

on:
  pull_request:
    branches: [ main ]
    paths:
    - 'backend/services/<name>/**'
    - 'backend/store/**'

jobs:
  build-carddeck:
    uses: ./.github/workflows/build_service.yml
    with:
      image-name: kioku_<name>
      path: ./backend/services/<name>
      context: ./backend
```

5. Add the service to the Kubernetes deployment
   1. Add the Kubernetes template in `helm/kioku/templates/<name>.yaml`
    ```yaml
    ---
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: "{{ .Values.<name>.name }}-deployment"
      labels:
        {{- include "kioku.<name>.labels" . | nindent 4 }}
    spec:
      replicas: 1
      selector:
        matchLabels:
          {{- include "kioku.<name>.labels" . | nindent 6 }}
      template:
        metadata:
          labels:
            {{- include "kioku.<name>.labels" . | nindent 8 }}
        spec:
          serviceAccountName: go-micro
          containers:
          - name: {{ .Values.<name>.name }}
            image: "{{ .Values.<name>.image }}:{{ .Values.<name>.tag }}"
          {{ if eq .Values.mode "production" }}
            imagePullPolicy: Always
          {{ else }}
            imagePullPolicy: Never
          {{ end }}
            ports:
              - containerPort: 8080
            resources:
              limits:
                cpu: 500m
                memory: 500M
              requests:
                cpu: 200m
                memory: 200M
            env:
              - name: HOSTNAME
                valueFrom:
                  fieldRef:
                    fieldPath: metadata.name
              - name: PORT
                value: "8080"
              - name: POSTGRES_PASSWORD
                valueFrom:
                  secretKeyRef:
                    name: {{ print "postgres." .Values.database.databaseName ".credentials.postgresql.acid.zalan.do" }}
                    key: password

            envFrom:
              - secretRef:
                  name: {{ .Values.database.secret.name }}
              - configMapRef:
                  name: service-env
    ---
    apiVersion: v1
    kind: Service
    metadata:
      name: "{{ .Values.<name>.name }}-service"
    spec:
      selector:
        {{- include "kioku.<name>.labels" . | nindent 4 }}
      ports:
        - port: 8080
          targetPort: 8080
    ---
    apiVersion: autoscaling/v1
    kind: HorizontalPodAutoscaler
    metadata:
    name: "hpa-{{ .Values.<name>.name }}-deployment"
    spec:
    scaleTargetRef:
      apiVersion: apps/v1
      kind: Deployment
      name: "{{ .Values.<name>.name }}-deployment"
    minReplicas: {{ .Values.<name>.autoscaler.min }}
    maxReplicas: {{ .Values.<name>.autoscaler.max }}
    targetCPUUtilizationPercentage: {{ .Values.<name>.autoscaler.targetCPUUtilizationPercentage }}
    ```
   2. Add configuration values in `helm/kioku/values.yaml`
    ```yaml
    carddeck:
    name: kioku-<name>
    image: ghcr.io/kioku-project/kioku_<name>
    tag: prod

    autoscaler:
      min: 1
      max: 10
      targetCPUUtilizationPercentage: 50
    ```
   3. Add labels `helm/kioku/templates/_helpers.tpl`
    ```yaml
    {{/*
    <name> labels
    */}}
    {{- define "kioku.<name>.labels" -}}
    app.kubernetes.io/name: {{ .Values.<name>.name }}
    {{- end }}
    ```

6. Adjust the proxy rules in the frontend service to be able to serve the new service if needed
```
